src/Network/Wai/Middleware/OpenApi/Validate.hs

```diff
@@ -0,0 +1,49 @@
+-- | Shared validation routine between request and response bodies
+module Network.Wai.Middleware.OpenApi.Validate
+  ( ValidateOptions (..)
+  , validate
+  ) where
+
+import Prelude
+
+import Control.Lens ((^.))
+import Data.Aeson (Value, decode)
+import Data.ByteString.Lazy qualified as BSL
+import Data.List.NonEmpty qualified as NE
+import Data.OpenApi qualified as OpenApi
+import Network.Wai (Request)
+import Network.Wai.Middleware.OpenApi.Settings
+import Network.Wai.Middleware.OpenApi.ValidationError
+
+data ValidateOptions extra a = ValidateOptions
+  { settings :: Settings
+  , request :: Request
+  , schema :: Maybe OpenApi.Schema
+  , getBody :: IO (BSL.ByteString, extra)
+  , respondActual :: Maybe extra -> IO a
+  , respondErrors :: ValidationErrors -> IO a
+  }
+
+-- | Interal DRY'ing of request-vs-response validation
+validate :: ValidateOptions extra a -> IO a
+validate options = do
+  case options.schema of
+    Nothing -> options.respondActual Nothing
```

Comment:
Could we do something here for a missing part of the schema?

Reply:
I don't know what you mean by "a missing part of the schema?". Can you expand?

Reply:
Doesn't a `Nothing` here imply that we hit a route that isn't in the spec?

Reply:
Oh yes. Sorry, I was confused when you said "a missing part of the schema"; thinking you meant a missing part of specifically the `Schema` being referenced on this line, and not "a missing part of the _spec_".

That does seem like something a strict validation shouldn't silently pass along -- but how should it behave? Does behavior depend on if the entire route is missing, vs the route being present but missing the method or response status definition?

Reply:
> Does behavior depend on if the entire route is missing, vs the route being present but missing the method or response status definition?

I don't think that distinction really matters too much but if you can easily distinguish those then it'd be nice for the error to have that detail in it.

Reply:
The behavior when the route isn't in the openapi spec at all should probably be configurable in settings; I think that could be a valuable tool toward getting gradual spec coverage.

1. Start writing routes, turn on validation only for the routes that are defined in the spec so far.

2. We think we've covered everything we want to cover, change the middleware settings to reject uncovered routes. If there are routes we explicitly don't want covered, conditional application of the middleware can be done by the application outside of this library.

Reply:
For more obscure things that a spec might be missing, e.g. schema for 503 status,

- is it fair to say that 100% OpenAPI coverage is a goal like an application can/should be working toward?
- Or is there a set of things that OpenAPI _can_ express but that an application may legitimately never care to describe?

Reply:
> behavior when the route isn't in the openapi spec at all should probably be configurable in settings

I think that's right. I started sketching a simple `strictMode: true`, but the issue is that once you do that, you can no longer assume `application/json`. The content-type requested, for which a schema was not found, could be anything at all. This means when we generate error responses, we now need to be _further_ configurable, or attempt to match the `content-type` of the response we we're validating / respect the `accept` header of the request we were validating. The complexity spirals really quickly.

> Or is there a set of things that OpenAPI can express but that an application may legitimately never care to describe?

I believe this is the case. I'm not going to say OpenAPI is _only useful_ for programmatic clients, but I'd say it is _mostly used_ for programmatic clients. Therefore, in the common case, it's entirely legitimate to elect not to describe anything that won't be experienced by such clients. Fancy-api has various non-JSON responses (auth, error pages, health-check, version), I see no value in spec'ing them. There are also responses generated by 3rd-party middleware. Even Servant apps, with fully generated OpenAPI, would lack those without extra work.

Reply:
I'm going to experiment with a different user-facing interface to see if it makes solving this problem easier. It's been a long path to get to what I have here currently, and I wouldn't say I love it.

Reply:
OK, the latest commit should make this better. I think the interface is simpler in general, and it should make it easier for users who want to be strict to implement things exactly as they might want. I tried to speak to it a bit in the new _Configuring_ section of the README.